### DISCUSSION TITLE
How to update form data in MongoDB via PUT method?

### DISCUSSION THREAD

**Discussion by K-Juno** [neuefische/questions#211] [Web] [answered]
Dear all,

I am trying to figure out how to update data that is already saved up in MongoDB by using (again) the form.

1. For starters, below code has had many alterations and this should be the root of my troubles:
 ```
import dbConnect from '../../../dbConnect';
import Lyrics from '../../../models/Lyrics';

export default async function updateHandler(req, res) {
  await dbConnect();

  if (req.method === 'PUT') {
    const { id } = req.query;

    await Lyrics.findByIdAndUpdate(id, req.body, { new: true });

    res.status(200).json({ message: 'Lyrics updated.' });
  }
}
```

2. I have also created new files for the whole PUT-action: (Update)Form component + [id] page.
```
  async function handleSubmit(event) {
    event.preventDefault();
    const form = event.target;
    const updatedTitle = form.title.value.trim();
    const updatedLyrics = form.lyrics.value.trim();

    const updatedSong = {
      id: id,
      title: updatedTitle,
      lyrics: updatedLyrics,
    };

    await fetch('/api/lyrics/update', {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(updatedSong),
    });

    onChangeLyrics(lyrics, updatedSong);
```

After an extensive research I only got so far and even got more confused by several statements about the issue of _id in Mongoose...??

Also, I am very curious about how to get the data out of the db back into the form and whether I need "defaultValue" for that.

My current PR: https://github.com/K-Juno/capstone-project/pull/13

Thanks so much in advance, everyone!!

**Comment by ThomasKirst** [comment, marked as answer]
Hi Juni,

This looks quite good so far. The only thing you would have to change is to make your route dynamic.

1. Rename your API route from `api/lyrics/update.js` to `api/lyrics/[id].js`
2. With this you can add a lyric id to your route, e.g.
`http://localhost:3000/api/lyrics/3hg3kj4hg3j4`
3. Your fetch from your form would then look like:
```js
// id is your lyrics id and it can be read in your api route as you did already
await fetch('/api/lyrics/' + id, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(updatedSong),
});
```

I hope this helps :)
– Thomas

**Reply by K-Juno** [reply]
Dear Thomas, first of all thanks so much for your reply and the solutions.
I already have an api/lyrics/[id].js file, which contains the removeHandler & the DELETE method, so I just added the updateHandler to the same page:

```
import dbConnect from '../../../dbConnect';
import Lyrics from '../../../models/Lyrics';

export default async function removeHandler(req, res) {
  await dbConnect();

  if (req.method === 'DELETE') {
    const { id } = req.query;

    await Lyrics.findByIdAndDelete(id);

    res.status(200).json({ message: 'Lyrics deleted.' });
  }
}

export async function updateHandler(req, res) {
  await dbConnect();

  if (req.method === 'PUT') {
    const { id } = req.query;

    await Lyrics.findByIdAndUpdate(id, req.body, { new: true });

    res.status(200).json({ message: 'Lyrics updated.' });
  }
}
```
My fetch looks like this now:
```
    await fetch('/api/lyrics/${id}', {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(updatedSong),
    });
```
But now I get a message:
`API resolved without sending a response for /api/lyrics/undefined, this may result in stalled requests.`
Help! :D

**Reply by ThomasKirst** [reply]
Hi Juni,

you are close :)

Your fetch should read:
```js
await fetch(`/api/lyrics/${id}`, { // backticks!
```

Your API route can only contain one function, e.g.:

```js
export default async function handler(req, res) {
  await dbConnect();

  if (req.method === 'DELETE') {
    const { id } = req.query;

    await Lyrics.findByIdAndDelete(id);

    res.status(200).json({ message: 'Lyrics deleted.' });
  }
  
  if (req.method === 'PUT') {
    const { id } = req.query;

    await Lyrics.findByIdAndUpdate(id, req.body, { new: true });

    res.status(200).json({ message: 'Lyrics updated.' });
  }
}
```

This should do the trick :)

**Reply by ThomasKirst** [reply]
For the next stage ('make it nice') you should consider a try/catch block and return something if neither DELETE or PUT is given.
